### PR TITLE
Change LOCK_global_table_stats to error-checking mutex to avoid deadlock

### PR DESCRIPTION
**Pull request by alxyang**
It's possible for this mutex to be locked twice by one thread when
`ha_myisam::write_row()` errors out during an information schema query.

We use an error checking mutex here so we can handle this situation.
Fixes https://github.com/facebook/mysql-5.6/issues/132.

### CODE REVIEW

File: mysql-test/include/check-testcase.test [hunk 1]
```diff
@@ -91,5 +91,5 @@ if (!$tmp) {
   SHOW SLAVE STATUS;
 }
 
-call mtr.check_testcase();
+#call mtr.check_testcase();
```
Comment (laurynas-biveinis) [inline mysql-test/include/check-testcase.test]: Is this change intentional?

File: storage/myisam/ha_myisam.cc [hunk 2]
```diff
@@ -813,6 +813,10 @@ int ha_myisam::write_row(uchar *buf)
 {
   ha_statistic_increment(&SSV::ha_write_count);
 
+  DBUG_EXECUTE_IF("myisam_simulate_crash_in_write_row", {
```
Comment (laurynas-biveinis) [inline storage/myisam/ha_myisam.cc]: "myisam_simulate_max_fs_exceeded_in_write_row" or similar?

Comment (alxyang) [reply storage/myisam/ha_myisam.cc]: that wouldn't be an entirely accurate name. this issue manifests for any error in write_row, as long as `fill_table_stats()` was called prior.  It's not necessary for max_fs to be exceeded.  Internally, we've run into this deadlock without that being the case.

though, "crash" doesn't seem like a good choice of words either, i'll rename it to `myisam_simulate_err_in_write_row`

File: mysql-test/include/check-testcase.test (hunk 1 above)
Comment (alxyang) [reply mysql-test/include/check-testcase.test]: nice catch.
